opentelemetry-sdk/src/attributes/set.rs

```diff
@@ -104,37 +105,46 @@ impl Eq for HashKeyValue {}
 ///
 /// This must implement [Hash], [PartialEq], and [Eq] so it may be used as
 /// HashMap keys and other de-duplication methods.
-#[derive(Clone, Default, Debug, Hash, PartialEq, Eq)]
-pub struct AttributeSet(BTreeSet<HashKeyValue>);
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct AttributeSet(Vec<HashKeyValue>, u64);
 
 impl From<&[KeyValue]> for AttributeSet {
     fn from(values: &[KeyValue]) -> Self {
-        let mut seen = HashSet::with_capacity(values.len());
-        AttributeSet(
-            values
-                .iter()
-                .rev()
-                .filter_map(|kv| {
-                    if seen.contains(&&kv.key) {
-                        None
-                    } else {
-                        seen.insert(&kv.key);
-                        Some(HashKeyValue(kv.clone()))
-                    }
-                })
-                .collect(),
-        )
+        let mut seen_keys = HashSet::with_capacity(values.len());
+        let mut vec = values
+            .iter()
+            .rev()
+            .filter_map(|k| {
+                if seen_keys.insert(k.key.clone()) {
+                    Some(HashKeyValue(k.clone()))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        vec.sort();
+
+        let mut hasher = DefaultHasher::new();
+        for value in &vec {
+            value.hash(&mut hasher);
+        }
+
+        AttributeSet(vec, hasher.finish())
     }
 }
 
 impl From<&Resource> for AttributeSet {
     fn from(values: &Resource) -> Self {
-        AttributeSet(
-            values
-                .iter()
-                .map(|(key, value)| HashKeyValue(KeyValue::new(key.clone(), value.clone())))
-                .collect(),
-        )
+        let vec = values
+            .iter()
+            .map(|(key, value)| HashKeyValue(KeyValue::new(key.clone(), value.clone())))
+            .collect::<Vec<_>>();
```

Comment:
Need to sort this vec as well now to ensure order is maintained

Reply:
Good catch, added!